### PR TITLE
[Trainer]Add  unified checkpoint for hybrid parallel.

### PR DESCRIPTION
**Pull request by ZHUI**
### PR types

New features

### PR changes

Trainer

### Description

Add unified checkpoint which merge tp tensor, and save all model weights to ` model.safetensors.index.json`.
- eg. for tp4pp2, 
- it will save `model-00001-of-00008.safetensors ~ model-00008-of-00008.safetensors`
- `00001-00004` store all tensor for pp00, `00005-00008` store all tensor for pp01
-  each of `00001-00004`  are tensor merge on tensor parallel, and equally part to 4 parts.

others: add `distributed_isfile`  and `distributed_file`  to copy some small files to other machine nodes.

TODO: 
- Support optimizer broadcast for dp.

#### Before
```python
-rw-rw-r-- 1 zhonghui03 zhonghui03       1011 Oct 16 15:34 config.json
-rw-rw-r-- 1 zhonghui03 zhonghui03 1138788706 Oct 16 15:34 model_state.tp00.pdparams
-rw-rw-r-- 1 zhonghui03 zhonghui03 1147194144 Oct 16 15:34 model_state.tp01.pdparams
-rw-rw-r-- 1 zhonghui03 zhonghui03 1124108982 Oct 16 15:34 model_state.tp02.pdparams
-rw-rw-r-- 1 zhonghui03 zhonghui03 1161882118 Oct 16 15:34 model_state.tp03.pdparams
```

#### After
```python
-rw-rw-r-- 1 zhonghui03 zhonghui03       1011 Oct 16 15:34 config.json
-rw-rw-r-- 1 zhonghui03 zhonghui03 1138788706 Oct 16 15:34 model-00001-of-00004.safetensors
-rw-rw-r-- 1 zhonghui03 zhonghui03 1147194144 Oct 16 15:34 model-00002-of-00004.safetensors
-rw-rw-r-- 1 zhonghui03 zhonghui03 1124108982 Oct 16 15:34 model-00003-of-00004.safetensors
-rw-rw-r-- 1 zhonghui03 zhonghui03 1161882118 Oct 16 15:34 model-00004-of-00004.safetensors
-rw-rw-r-- 1 zhonghui03 zhonghui03       7986 Oct 16 15:34 model.safetensors.index.json
```

```json
{
    "metadata": {
        "total_size": 4571963392
    },
    "weight_map": {
        "llama.embed_tokens.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.self_attn.q_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.self_attn.k_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.self_attn.v_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.self_attn.o_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.mlp.gate_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.mlp.up_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.mlp.down_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.input_layernorm.weight": "model-00001-of-00004.safetensors",
        "llama.layers.0.post_attention_layernorm.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.self_attn.q_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.self_attn.k_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.self_attn.v_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.self_attn.o_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.mlp.gate_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.mlp.up_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.mlp.down_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.input_layernorm.weight": "model-00001-of-00004.safetensors",
        "llama.layers.1.post_attention_layernorm.weight": "model-00001-of-00004.safetensors",
        "llama.layers.2.self_attn.q_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.2.self_attn.k_proj.weight": "model-00001-of-00004.safetensors",
        "llama.layers.2.self_attn.v_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.2.self_attn.o_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.2.mlp.gate_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.2.mlp.up_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.2.mlp.down_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.2.input_layernorm.weight": "model-00002-of-00004.safetensors",
        "llama.layers.2.post_attention_layernorm.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.self_attn.q_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.self_attn.k_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.self_attn.v_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.self_attn.o_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.mlp.gate_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.mlp.up_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.mlp.down_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.input_layernorm.weight": "model-00002-of-00004.safetensors",
        "llama.layers.3.post_attention_layernorm.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.self_attn.q_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.self_attn.k_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.self_attn.v_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.self_attn.o_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.mlp.gate_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.mlp.up_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.mlp.down_proj.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.input_layernorm.weight": "model-00002-of-00004.safetensors",
        "llama.layers.4.post_attention_layernorm.weight": "model-00002-of-00004.safetensors",
        "llama.layers.5.self_attn.q_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.self_attn.k_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.self_attn.v_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.self_attn.o_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.mlp.gate_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.mlp.up_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.mlp.down_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.input_layernorm.weight": "model-00003-of-00004.safetensors",
        "llama.layers.5.post_attention_layernorm.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.self_attn.q_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.self_attn.k_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.self_attn.v_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.self_attn.o_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.mlp.gate_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.mlp.up_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.mlp.down_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.input_layernorm.weight": "model-00003-of-00004.safetensors",
        "llama.layers.6.post_attention_layernorm.weight": "model-00003-of-00004.safetensors",
        "llama.layers.7.self_attn.q_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.7.self_attn.k_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.7.self_attn.v_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.7.self_attn.o_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.7.mlp.gate_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.7.mlp.up_proj.weight": "model-00003-of-00004.safetensors",
        "llama.layers.7.mlp.down_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.7.input_layernorm.weight": "model-00004-of-00004.safetensors",
        "llama.layers.7.post_attention_layernorm.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.self_attn.q_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.self_attn.k_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.self_attn.v_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.self_attn.o_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.mlp.gate_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.mlp.up_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.mlp.down_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.input_layernorm.weight": "model-00004-of-00004.safetensors",
        "llama.layers.8.post_attention_layernorm.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.self_attn.q_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.self_attn.k_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.self_attn.v_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.self_attn.o_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.mlp.gate_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.mlp.up_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.mlp.down_proj.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.input_layernorm.weight": "model-00004-of-00004.safetensors",
        "llama.layers.9.post_attention_layernorm.weight": "model-00004-of-00004.safetensors",
        "llama.norm.weight": "model-00004-of-00004.safetensors",
        "lm_head.weight": "model-00004-of-00004.safetensors"
    }
}
```

### CODE REVIEW

**Comment by ZHUI** [comment]
from https://github.com/PaddlePaddle/PaddleNLP/pull/7143

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    """_summary_
+
+    Args:
+        args (_type_): _description_
+        model (_type_): model to save
+        output_dir (_type_): save dir
+        safe_serialization (bool, optional): use safetensors. Defaults to False.
+
+    Raises:
+        ValueError: if model is an instanceof `PretrainedModel` and the model cannot be saved
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: 这个raises少了个not

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        if not safe_serialization:
+            path = os.path.join(output_dir, PADDLE_WEIGHTS_INDEX_NAME)
+        else:
+            path = os.path.join(output_dir, SAFE_WEIGHTS_INDEX_NAME)
+
+        with open(path, "w") as f:
+            json.dump(sharded_index, f, indent=4)
+
+
+def load_unified_checkpoint(model: paddle.nn.Layer, resume_from_checkpoint: str, safe_serialization=False) -> None:
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: paddle.nn.Layer是不是改成PretrainedModel更好？

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    """Load potential model checkpoint
+
+    Args:
+        model (nn.Layer): Your model to save
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: Your model to load

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    safe_serialization=False,
+):
+    """Get state_dict and config to save
+
+    Args:
+        model_to_save (nn.Layer): model to, save
+        safe_serialization (bool, optional): safe serialization using safetensors. Defaults to False.
+
+    Returns:
+        tuple: state_dict, and  config
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: 注释return的东西少了

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+
+
+def filter_params(model_to_save, state_dict):
+    hcg = fleet.get_hybrid_communicate_group()
+    tp_group = hcg.get_model_parallel_group()
+
+    tp_size = tp_group.nranks
+    tp_rank = tp_group.rank
+
+    filter_tensor_list = [[] for i in range(tp_size)]
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:297]: 这行删掉

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+                    filter_tensor_list.append(current_block)
+                current_block = []
+                current_block_size = 0
+
+            current_block.append(key)
+            current_block_size += weight_size
+            total_size += weight_size
+
+        filter_tensor_list.append(current_block)
+        assert len(filter_tensor_list) == tp_size, "Error, partion failed!"
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: partition

File: paddlenlp/trainer/trainer.py [hunk 7]
```diff
@@ -495,6 +498,16 @@ def _load_from_checkpoint(self, resume_from_checkpoint=None):
...
+            if resume_from_checkpoint is not None and self.args.dataset_rank == 0:
+                load_unified_checkpoint(
+                    self.model,
+                    resume_from_checkpoint,
+                    safe_serialization=True,
+                )
+                logger.info(f"Loading model from {resume_from_checkpoint} using unified checkpoint.")
+                return
+
         if isinstance(self.model, LoRAModel) or isinstance(self.model, PrefixModelForCausalLM):
```
Comment (wawltor) [inline paddlenlp/trainer/trainer.py:511]: 这里对lora model 和 p-tuning的热启训练看起来没有衔接，这里最好记个TODO吧

File: paddlenlp/trainer/trainer.py [hunk 8]
```diff
@@ -2040,15 +2057,24 @@ def _rotate_checkpoints(self, use_mtime=False, output_dir=None) -> None:
...
-        merge_tensor_parallel = merge_tensor_parallel and self.args.use_hybrid_parallel
+        if self.args.should_save:
+            if self.tokenizer is not None:
+                self.tokenizer.save_pretrained(output_dir)
+            # Good practice: save your training arguments together with the trained model
+            paddle.save(self.args, os.path.join(output_dir, TRAINING_ARGS_NAME))
+
+        if self.args.unified_checkpoint:
+            save_unified_checkpoint(self.args, self.model, output_dir, safe_serialization=True)
+            return
```
Comment (wawltor) [inline paddlenlp/trainer/trainer.py:2074]: 和Load模型一样，没有适配LoRA和P-tuning模型

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    dtype = get_parameter_dtype(model_to_save)
+    model_to_save.config.dtype = str(dtype).split(".")[1]
+    config_to_save = copy.deepcopy(model_to_save.config)
+
+    if config_to_save.tensor_parallel_degree > 1:
+        state_dict = merge_tensor_parallel_with_shard(model_to_save, state_dict, config_to_save, all_filter_keys)
+
+    if config_to_save.tensor_parallel_degree > 1:
+        # do we need to change?
+        config_to_save.tensor_parallel_degree = 1
```
Comment (wawltor) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:217]: 这里的改动为什么了？那热启训练的时候这个会有影响吗？因为这个值是从外部传入的

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    from safetensors.numpy import save_file as safe_save_file
+
+
+__all__ = [
+    "load_unified_checkpoint",
+    "save_unified_checkpoint",
+]
+
+
+def save_unified_checkpoint(args, model, output_dir, safe_serialization=False):
```
Comment (wawltor) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:59]: 这个函数的调用在哪里会有限制只是在一路DP里面限制了？

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+                tensor_bytes_dict[k] = v.numel().item() * tp_size * dtype_byte_size(v.dtype)
+            else:
+                tensor_bytes_dict[k] = v.numel().item() * dtype_byte_size(v.dtype)
+
+        filter_tensor_list = []
+        current_block = []
+        current_block_size = 0
+        total_size = 0
+
+        max_shard_size = (sum(tensor_bytes_dict.values()) + tp_size - 1) // tp_size
```
Comment (wawltor) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:322]: 这里不需要考虑pipeline的模式吗？

File: paddlenlp/trainer/plugins/unified_checkpoint.py (hunk 10 above)
Comment (DesmonDay) [reply paddlenlp/trainer/plugins/unified_checkpoint.py:59]: 已线下回复

File: paddlenlp/trainer/plugins/unified_checkpoint.py (hunk 11 above)
Comment (DesmonDay) [reply paddlenlp/trainer/plugins/unified_checkpoint.py:322]: 不需要

File: paddlenlp/trainer/plugins/unified_checkpoint.py (hunk 10 above)
Comment (ZHUI) [reply paddlenlp/trainer/plugins/unified_checkpoint.py:59]: https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/trainer/trainer.py#L1891-L1892

**Review by wawltor — APPROVED** [review]
LGTM

**Pull request merged**